kafka-client/src/main/java/dev/responsive/kafka/clients/OffsetRecorder.java

```diff
@@ -0,0 +1,163 @@
+package dev.responsive.kafka.clients;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * Record offsets written to kafka and committed to consumer groups. When a commit happens,
+ * notify listeners about the set of offsets.
+ *
+ * The callbacks here need to be synchronized. Almost all the callbacks happen on the stream
+ * thread. The lone exception is the Send callback, which happens on Producer I/O threads.
+ */
+class OffsetRecorder {
+  private final Map<RecordingKey, Long> uncommitted = new HashMap<>();
+  private final Map<TopicPartition, Long> written = new HashMap<>();
+  private final ProducerListener producerListener = new ProducerListener();
+  private final ConsumerListener consumerListener = new ConsumerListener();
+  private final List<CommitCallback> commitCallback = new LinkedList<>();
+  private final boolean eos;
+
+  OffsetRecorder(final boolean eos) {
+    this.eos = eos;
+  }
+
+  public synchronized void addCommitCallback(final CommitCallback callback) {
+    commitCallback.add(callback);
+  }
+
+  public synchronized void recordConsumedOffsets(
+      final String consumerGroup,
+      final TopicPartition partition,
+      final long offset
+  ) {
+    uncommitted.put(new RecordingKey(partition, consumerGroup), offset);
+  }
+
+  public synchronized void recordProduce(final RecordMetadata recordMetadata) {
```

Comment:
```suggestion
  public synchronized void onProduced(final RecordMetadata recordMetadata) {
```
(same for `recordX` below) at first I was confused because I thought this produces records!